pmlc/dialect/stdx/transforms/subgroup_block_ops.cc

```diff
@@ -0,0 +1,176 @@
+///    store %1, %arg1[%0] : memref<64xf32>
+///    scf.yield
+///  }
+///
+/// Code after transformation:
+///
+///  %c0 = constant 0 : index
+///  %c1 = constant 1 : index
+///  %c8 = constant 8 : index
+///  %c64 = constant 64 : index
+///  scf.parallel (%i, %sid) = (%c0, %c0) to (%c64, %c8) step (%c8, %c1) {
+///    %idx = addi %i, %sid : index
+///    %0 = stdx.subgroup_block_read_intel %arg0[%i] : memref<64xf32>
+///    stdx.subgroup_block_write_intel %0, %arg1[%i] : memref<64xf32>
+///    scf.yield
+
+class BlockOpsImpl {
+private:
+  scf::ParallelOp loop;
+  DenseSet<Operation *> toReplace;
+
+  template <typename OpType>
+  void checkLoadStore(OpType loadOp) {
+    // The loadOp/storeOp needs to be stride 1,
+    // and the last index needs to be from AddI operation with subgroup_id
+    // (this is modelled in subgroup_broadcast pass)
+    //
+    // TODO: Verify if this condition is correct, or it
+    // will be needed to check for the additional parallel
+    // op gpu thread attribute
+    if (auto addIOp =
+            dyn_cast<AddIOp>(loadOp.indices().back().getDefiningOp())) {
+      auto addIOperand = addIOp.getOperand(1);
+      auto lastIndiceOfLoop = loop.getBody()->getArguments().back();
+      int64_t memRefOffset;
+      SmallVector<int64_t, 4> memRefStrides;
+      getStridesAndOffset(loadOp.getMemRefType(), memRefStrides, memRefOffset);
+      if ((addIOperand == lastIndiceOfLoop) && (memRefStrides.back() == 1))
+        toReplace.insert(loadOp);
+    }
+  }
+
+public:
+  explicit BlockOpsImpl(scf::ParallelOp loop) : loop(loop) {}
+
+  void replaceLoadOp(LoadOp op) {
+    // Replace with SubgroupBlockWriteINTELOp, update indices to remove
+    // subgroup_id, we know at this point that last index comes from the
+    // AddI operation, where second operand is subgroup_id
+    OpBuilder builder(op);
+    SmallVector<Value, 4> newIndices = op.indices();
+    auto addOp = newIndices.back().getDefiningOp();
+    newIndices.pop_back();
+    newIndices.push_back(addOp->getOperand(0));
+    auto newBlockReadOp = builder.create<SubgroupBlockReadINTELOp>(
+        op.getLoc(), op.memref(), newIndices);
+    op.replaceAllUsesWith(newBlockReadOp.getResult());
+    op.erase();
+  }
+
+  void replaceStoreOp(StoreOp op) {
+    // Replace with SubgroupBlockWriteINTELOp, update indice to remove
+    // subgroup_id, we know at this point that last indice comes from the
+    // addi operation, where second operand is subgroup_id
+    OpBuilder builder(op);
+    SmallVector<Value, 4> newIndices = op.indices();
+    auto addOp = newIndices.back().getDefiningOp();
+    newIndices.pop_back();
+    newIndices.push_back(addOp->getOperand(0));
+    builder.create<SubgroupBlockWriteINTELOp>(op.getLoc(), op.value(),
+                                              op.memref(), newIndices);
+    op.erase();
+  }
+
+  LogicalResult blockOps() {
+    mlir::Block *body = loop.getBody();
+
+    for (auto &op : body->getOperations()) {
```

Comment:
There seems to be no reason to preflight all the ops, and then replace them, rather than verify and replace them as you go.  Additionally, the replaceLoadOp/replaceStoreOp functions do not check toReplace (which is incorrect, and will replace all loads!).  It seems easier to do something like:

for (auto& loadOp : body->getOps<LoadOp>()) {
  if (succeeded(checkLoadStore(loadOp))
    replaceLoadOp(loadOp);
}

And similar for store.

Reply:
yes, that was a mistake after code rework, it shoul replace only the ops from toReplace list, not all of them, anyways changed this to single loop now as requested